File: uwnav/io/timebase.py
# uwnav/io/timebase.py
# -*- coding: utf-8 -*-
"""
统一时间基（Python 版）

与 C++ 版 uwnav::timebase 对齐的最小实现：

- 仅使用单调时钟 time.monotonic_ns() 作为唯一时间源
- 不再维护 real/epoch 语义，也不再做“真实墙上时间”估计
- 提供按传感器类型区分的默认延迟（LatencyDefaults）
- 提供统一时间戳结构 Stamp：
    * host_time_ns      —— 采集/解析瞬间的主机单调时间
    * corrected_time_ns —— host_time_ns - latency_ns
    * sensor_time_ns    —— 可选，预留给将来使用传感器内部时间戳
    * latency_ns        —— 本次使用的延迟估计

推荐用法（采集脚本中）：
    from uwnav.io.timebase import SensorKind, stamp

    ts = stamp("imu0", SensorKind.IMU)
    packet.mono_ns = ts.host_time_ns
    packet.est_ns  = ts.corrected_time_ns
"""

from __future__ import annotations

import enum
import time
from dataclasses import dataclass
from typing import Optional

__all__ = [
    "SensorKind",
    "Stamp",
    "LatencyDefaults",
    "latency_defaults",
    "now_ns",
    "stamp",
    "to_sec",
]


# ========================= 基础时间源 =========================


def now_ns() -> int:
    """
    返回当前单调时间（ns）。
    对应 C++ 的 uwnav::timebase::now_ns()
    """
    return time.monotonic_ns()


# ========================= 传感器类型 & 时间戳结构 =========================


class SensorKind(enum.Enum):
    IMU = "imu"
    DVL = "dvl"
    USBL = "usbl"
    OTHER = "other"


@dataclass
class Stamp:
    """
    与 C++ 版 Stamp 对齐的 Python 结构：
    - sensor_id: 传感器实例标识，例如 "imu0", "dvl0"
    - kind:      传感器类型（IMU/DVL/USBL/OTHER）
    - host_time_ns:      调用 stamp() 时的主机单调时间
    - corrected_time_ns: host_time_ns - latency_ns
    - sensor_time_ns:    可选，预留将来用传感器内部时间戳对齐
    - latency_ns:        本次使用的延迟估计（根据 kind 或显式传入）
    """

    sensor_id: str
    kind: SensorKind
    host_time_ns: int
    corrected_time_ns: int
    sensor_time_ns: Optional[int] = None
    latency_ns: int = 0


# ========================= 默认延迟配置 =========================


@dataclass
class LatencyDefaults:
    """
    默认延迟（单位：ns）

    注意：这些值是工程经验参数，需在下水联调时根据实际情况微调：
      - imu_ns:   IMU 从测量到到达主机的典型时延
      - dvl_ns:   DVL 帧从测量到到达主机的典型时延
      - usbl_ns:  USBL/其它慢速定位传感器的典型时延
      - other_ns: 其它未分类传感器的默认时延
    """
    imu_ns: int = 2_000_000        # 2 ms
    dvl_ns: int = 50_000_000       # 50 ms
    usbl_ns: int = 150_000_000     # 150 ms
    other_ns: int = 0              # 默认不补偿


_latency_defaults = LatencyDefaults()


def latency_defaults() -> LatencyDefaults:
    """
    获取可修改的默认延迟配置对象。

    示例：
        from uwnav.io.timebase import latency_defaults
        cfg = latency_defaults()
        cfg.dvl_ns = 80_000_000
    """
    return _latency_defaults


def _default_latency_ns(kind: SensorKind) -> int:
    d = _latency_defaults
    if kind is SensorKind.IMU:
        return d.imu_ns
    if kind is SensorKind.DVL:
        return d.dvl_ns
    if kind is SensorKind.USBL:
        return d.usbl_ns
    return d.other_ns


# ========================= 核心 API：stamp =========================


def stamp(
    sensor_id: str,
    kind: SensorKind,
    sensor_time_ns: Optional[int] = None,
    latency_ns: Optional[int] = None,
) -> Stamp:
    """
    统一打时间戳的核心函数（Python 版）。

    参数：
        sensor_id:      传感器实例 ID，如 "imu0", "dvl0"
        kind:           传感器类型（SensorKind.IMU/DVL/USBL/OTHER）
        sensor_time_ns: 可选，传感器内部时间（将来做更精细对齐时使用）
        latency_ns:     可选，显式指定延迟；若为 None，则使用默认值

    返回：
        Stamp 对象，包含：
            host_time_ns
            corrected_time_ns
            sensor_time_ns
            latency_ns

    当前实现：
        host_time_ns      = time.monotonic_ns()
        latency_ns        = 显式传入 或 默认延迟
        corrected_time_ns = host_time_ns - latency_ns
    """
    host = now_ns()
    lat = latency_ns if latency_ns is not None else _default_latency_ns(kind)
    corrected = host - lat

    return Stamp(
        sensor_id=sensor_id,
        kind=kind,
        host_time_ns=host,
        corrected_time_ns=corrected,
        sensor_time_ns=sensor_time_ns,
        latency_ns=lat,
    )


# ========================= 工具函数 =========================


def to_sec(ns: int) -> float:
    """
    ns → 秒（float）
    """
    return ns / 1e9

File: nav_core/src/timebase.cpp
// nav_core/src/timebase.cpp
//
// 时间基准与统一时间戳工具实现：uwnav::core::timebase
//
// 设计目标：
//   - 所有传感器（IMU / DVL / USBL / 其他）统一走 steady_clock 时间基；
//   - 通过一个简单的 Stamp 结构，把“主机时间 + 估计延迟”封装为 corrected_time_ns；
//   - 后端 ESKF / Logger / 跨进程 NavStatePublisher 都可以只认一个时间轴；
//   - 将来可以在这里扩展“传感器自带时间戳”的校准逻辑，而不影响上层代码。

#include "nav_core/timebase.hpp"

namespace uwnav::core::timebase {

// ============================================================================
// 基础时间接口
// ============================================================================

MonoTimeNs now_ns()
{
    const auto tp = Clock::now().time_since_epoch();
    return std::chrono::duration_cast<std::chrono::nanoseconds>(tp).count();
}

TimePoint now()
{
    return Clock::now();
}

// ============================================================================
// 默认延迟（全局静态配置）
// ============================================================================
//
// 说明：
//   - 这里的延迟是“从物理世界发生”到“主机收到并解析完成”的估计值；
//   - 当前版本先写死一个 LatencyDefaults 全局实例，后续可以从 YAML / config 加载；
//   - 导航/融合模块只看 corrected_time_ns，不关心内部怎么算。
// ============================================================================

static LatencyDefaults g_latency_defaults{};

LatencyDefaults& latency_defaults()
{
    return g_latency_defaults;
}

// 根据传感器类型，返回默认延迟（单位：ns）
static std::int64_t default_latency_ns(SensorKind kind)
{
    const auto& d = g_latency_defaults;
    switch (kind) {
    case SensorKind::IMU:  return d.imu_ns;
    case SensorKind::DVL:  return d.dvl_ns;
    case SensorKind::USBL: return d.usbl_ns;
    default:               return d.other_ns;
    }
}

// ============================================================================
// stamp(): 统一时间戳核心方法
// ============================================================================
//
// 使用方式（以 IMU 为例）：
//   auto ts = uwnav::timebase::stamp("imu0", SensorKind::IMU);
//   frame.mono_ns = ts.host_time_ns;
//   frame.est_ns  = ts.corrected_time_ns;
//
// 说明：
//   - host_time_ns        ：调用时刻的 steady_clock 纳秒值（主机时间）
//   - sensor_time_ns(opt) ：可选的“传感器内部时间”（目前仅存放，不做对齐）
//   - latency_ns          ：估计延迟；不传则用 default_latency_ns(kind)
//   - corrected_time_ns   ：host_time_ns - latency_ns
//
// 将来如果需要做“传感器时钟对齐”（例如 DVL/USBL 自带时间），可以在这里扩展：
//   - 利用 sensor_time_ns 做 offset 估计；
//   - 形成 host_time_ns <-> sensor_time_ns 的映射；
//   - 再统一写入 corrected_time_ns。
//
Stamp stamp(const std::string&              sensor_id,
            SensorKind                      kind,
            std::optional<MonoTimeNs>       sensor_time_ns,
            std::optional<std::int64_t>     latency_ns)
{
    Stamp s;
    s.sensor_id      = sensor_id;
    s.kind           = kind;
    s.host_time_ns   = now_ns();
    s.sensor_time_ns = sensor_time_ns;

    // 1) 选择延迟：调用方显式传入优先，否则使用默认值
    s.latency_ns = latency_ns.has_value()
        ? *latency_ns
        : default_latency_ns(kind);

    // 2) 统一后的 corrected 时间戳（当前版本 = host_time_ns - latency）
    s.corrected_time_ns = s.host_time_ns - s.latency_ns;

    // TODO（将来）：如需利用 sensor_time_ns 做时钟对齐，在这里扩展逻辑
    // if (sensor_time_ns.has_value()) {
    //     // 估计传感器时钟与 host_time_ns 的偏移，修正 corrected_time_ns
    // }

    return s;
}

} // namespace uwnav::core::timebase
